functions/src/index.ts

```diff
@@ -65,28 +109,67 @@ exports.thingspeakToFirestore = functions
       );
 
       const resolvedPath = readingsSubcollection(knownSensor.id);
-
-      // Only add data if not already present in database.
-      // This happens if a sensor is down, so only old data is returned.
       const readingsRef = firestore.collection(resolvedPath);
+
+      // If data is not already present in the database add
+      // it to the historical readings
+      let firestoreSafeReading = defaultPm25BufferElement;
       if (
         (
           await readingsRef
             .where('timestamp', '==', Timestamp.fromDate(reading.timestamp))
             .get()
         ).empty
       ) {
-        const firestoreSafeReading = {
+        // Update reading with values
+        firestoreSafeReading = {
           timestamp: Timestamp.fromDate(reading.timestamp),
           channelAPm25: reading.channelAPm25,
           channelBPm25: reading.channelBPm25,
           humidity: reading.humidity,
           latitude: reading.latitude,
           longitude: reading.longitude,
         };
+        // Add to historical readings
         await readingsRef.add(firestoreSafeReading);
       }
 
+      // Add readings to the pm25 buffer
+      const docRef = firestore.collection('sensors').doc(knownSensor.id);
+
+      await docRef.get().then(doc => {
+        if (doc.exists) {
+          const status = doc.get('pm25BufferStatus');
+          if (status === bufferStatus.Exists) {
+            // If the buffer exists, update normally
+            let pm25BufferIndex = doc.get('pm25BufferIndex');
+            const pm25Buffer = doc.get('pm25Buffer');
+            // Add to circular buffer
+            pm25Buffer[pm25BufferIndex] = firestoreSafeReading;
+            // Increment index
```

Comment:
This comment feels like it repeats the name of the function call and might not be necessary, but that's up to debate